### PR TITLE
Collect all dependency related exceptions from all plugins and give them

### LINKED ISSUE #297: Collect dependency-related exceptions from all plugins before exiting
Currently the first unresolvable dependency leads to an exception in https://github.com/jenkinsci/plugin-installation-manager-tool/blob/c72ad4a42f6cdbe70615e8a322442b6da0c6b8ce/plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java#L1008 which exits the process.
To make fixing easier all exceptions should be collected from the plugin list and only then all of them should be emitted and the tool should exit.

### PR DESCRIPTION
**Pull request by kwin**
out as batch

This closes #297

### CODE REVIEW

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/AggregatePluginPrerequisitesNotMetException.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package io.jenkins.tools.pluginmanager.impl;
+
+import java.util.Collection;
+import java.util.stream.Collectors;
+
+public class AggregatePluginPrerequisitesNotMetException extends RuntimeException {
+
+    public AggregatePluginPrerequisitesNotMetException(Collection<Exception> exceptions) {
+        super("Multiple plugin prerequisites not met:\n" + exceptions.stream().map(Exception::getMessage).collect(Collectors.joining(",\n")));
```
Comment (timja) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/AggregatePluginPrerequisitesNotMetException.java:9]: could this get some test coverage please? and also manual steps to reproduce / test it?

Comment (kwin) [reply plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/AggregatePluginPrerequisitesNotMetException.java:9]: Done in https://github.com/jenkinsci/plugin-installation-manager-tool/pull/298/commits/c3b5e06eb24244407dd38aa8f60125f241faa9d4

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java [hunk 2]
```diff
@@ -185,15 +185,17 @@ public void start(boolean downloadUc) {
...
         pluginsToBeDownloaded = findPluginsToDownload(allPluginsAndDependencies);
         effectivePlugins = findEffectivePlugins(pluginsToBeDownloaded);
 
         listPlugins();
         showSpecificSecurityWarnings(pluginsToBeDownloaded);
-        checkVersionCompatibility(jenkinsVersion, pluginsToBeDownloaded);
-
+        checkVersionCompatibility(jenkinsVersion, pluginsToBeDownloaded, exceptions);
+        if (!exceptions.isEmpty()) {
+            throw new AggregatePluginPrerequisitesNotMetException(exceptions);
```
Comment (timja) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java:197]: can we only output unique results?

```
➜  plugin-installation-manager-tool git:(feature/batch-prerequisites-not-met-exceptions) java -jar plugin-management-cli/target/jenkins-plugin-manager-2.7.2-SNAPSHOT.jar --plugins blueocean --plugins branch-api:2.4.0 -d plugins/ --war ~/Downloads/jenkins.war --credentials abcd:acd:acd
Multiple plugin prerequisites not met:
Plugin blueocean:1.24.4 depends on branch-api:2.6.3, but there is an older version defined on the top level - branch-api:2.4.0,
Plugin blueocean:1.24.4 depends on branch-api:2.6.3, but there is an older version defined on the top level - branch-api:2.4.0,
Plugin blueocean:1.24.4 depends on branch-api:2.6.3, but there is an older version defined on the top level - branch-api:2.4.0,
Plugin blueocean:1.24.4 depends on branch-api:2.6.3, but there is an older version defined on the top level - branch-api:2.4.0
```

Comment (kwin) [reply plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java:197]: @timja How I can I reproduce? Which war have you used? This rather indicates that the algorithm should be optimized. Filtering in the exception would hide that flaw IMHO....

Comment (timja) [reply plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java:197]: it's jenkins-2.273, https://github.com/jenkinsci/jenkins/releases/download/jenkins-2.273/jenkins.war

Comment (kwin) [reply plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java:197]: With https://github.com/jenkinsci/plugin-installation-manager-tool/pull/298/commits/d2a74719c92584d3b33c98bf86af3bee4e8dc3fd you get
```
Multiple plugin prerequisites not met:
Plugin blueocean:1.24.4 (transitively via blueocean-pipeline-api-impl:1.24.4) depends on branch-api:2.6.3, but there is an older version defined on the top level - branch-api:2.4.0,
Plugin blueocean:1.24.4 (transitively via blueocean-autofavorite:1.2.4) depends on branch-api:2.6.3, but there is an older version defined on the top level - branch-api:2.4.0,
Plugin blueocean:1.24.4 (transitively via blueocean-pipeline-scm-api:1.24.4) depends on branch-api:2.6.3, but there is an older version defined on the top level - branch-api:2.4.0,
Plugin blueocean:1.24.4 (transitively via workflow-multibranch:2.22) depends on branch-api:2.6.3, but there is an older version defined on the top level - branch-api:2.4.0
```

That clarifies that this version is requested from 4 different transitive dependencies.
I would not consolidate to not hide this information from the called.

**Review by timja — APPROVED** [review]

**Pull request merged**